[WIP] Child blocks

noisysocks opened the pull request:
### What this is 🧐

WIP attempt at implementing https://github.com/WordPress/gutenberg/issues/5540. 

The meat of these changes is a new `allowedParents` property which lets you specify which blocks a child is allowed to be nested into:

```js
registerBlockType( 'acme/buy-button', {
	title: 'Buy Button',
	icon: 'cart',
	category: 'common',

	allowedParents: [ 'acme/product' ],
	...
} );
```

See the [test plugin](https://gist.github.com/noisysocks/9b9503bd6489ffa51088d35c7a2a8ba0) for a full code example. I reccomend installing the plugin, checking out this branch, and playing around with the API.

### Implementation details 🤓

#### Block registration API changes

- Introduces `allowedParents` which lets one specify which block types and post types a block can be nested into.
- Deprecates `isPrivate`. Instead, use `allowedParents: false`.

#### Selector changes

- `getInserterItems` now respects the above properties and is the source of truth for whether a block is insertable or not.
- Items returned by `getInserterItems` will include new `utility` and `frecency` attributes. `utility` indicates how useful we think inserting the block will be. This lets us show contextually useful blocks at the top of the Suggested tab.
- Deprecates `getSupportedBlockTypes`. `getInserterItems` will remove any blocks that are not supported.
- Deprecates `getFrecentInserterItems`. Instead, call `getInserterItems` and filter/sort by `frecency`.

#### Inserter changes

- Both `inserter` and `inserter-with-shortcuts` have been changed to use `getInserterItems` and its new functionality.

### How I've tested this ✅

I wrote a [little test plugin](https://gist.github.com/noisysocks/9b9503bd6489ffa51088d35c7a2a8ba0) which defines a custom post type (_Store_) and two blocks (_Product_ & _Buy Button_).

If you install the plugin, you can see that:

1. _Product_ can only be inserted into a _Store_ CPT.
2.  _Buy Button_ can only be inserted into a _Product_.
    1. This occurs even when `allowedBlocks` does not contain _Buy Button_, because `allowedParents` takes precedence.
3. When inserting _Buy Button_, the block **always** appears at the top of the inserter.
    1. This occurs even when `allowedBlocks` is set to `true`.
4. The inserter behaves like it always has.

### Outstanding questions 🤔

1. Is this API flexible enough for plugin developers?
3. Should `getInserterItems` be a selector?  
  **Pro:** It's technically a pure function that takes state as an input.  
  **Pro:** A selector means that this API is available to plugin developers.  
  **Con:** It's *way* more complicated than a regular selector.

### Still to do 👨‍💻

- [x] Write a test plugin
- [x] Fix unit tests for `getInserterItems`
- [x] Improve `getInserterItems` performance (e.g. by using `createSelector`)
- [ ] Write documentation for new API

noisysocks commented:
@gziolo @mtias What do you think of this API for Child Blocks? Check out [this gist](https://gist.github.com/noisysocks/9b9503bd6489ffa51088d35c7a2a8ba0) for an example of a plugin that uses it.

jasonbahl commented:
This looks awesome! This currently missing functionality is a pretty big blocker (pun intended) for using Gutenberg in production projects.

noisysocks commented:
I've updated this PR to use the API I described in https://github.com/WordPress/gutenberg/issues/5540#issuecomment-386192518. Check out [this gist](https://gist.github.com/noisysocks/9b9503bd6489ffa51088d35c7a2a8ba0) for an example usage. I think it works quite nicely. Any thoughts before I add documentation and tests?

@gziolo @mtias @jorgefilipecosta

gziolo commented on editor/store/selectors.js:
```diff
@@ -1584,32 +1615,3 @@ export function inSomeHistory( state, predicate ) {
- *
- * @return {string[]|boolean} Blocks that can be nested inside the block with the specified uid, or true/false to enable/disable all types.
- */
-export function getSupportedBlocks( state, uid, globallyEnabledBlockTypes ) {
```
Should it also be deprecated for some time?

noisysocks replied:
Yes—good catch 😄

gziolo reviewed (commented):
I think that there is a general agreement that it should be implemented as you proposed even though it might be confusing in some cases. On the technical side I like where this is going and would love to see it polished.

noisysocks commented:
Thanks @gziolo. I'll get this ready to ship!

noisysocks committed:
Support Child Blocks

Block registration API changes:

- Introduces `allowedParents` which lets one specify which block types
  and post types a block can be nested into.
- Deprecates `isPrivate`. Instead, use `allowedParents: false`.

Selector changes:

- `getInserterItems` now respects the above properties and is the source
  of truth for whether a block is insertable or not.
- Items returned by `getInserterItems` will include new `utility` and
  `frecency` attributes. `utility` indicates how useful we think inserting
  the block will be. This lets us show contextually useful blocks at the
  top of the Suggested tab.
- Deprecates `getSupportedBlockTypes`. `getInserterItems` will remove
  any blocks that are not supported.
- Deprecates `getFrecentInserterItems`. Instead, call `getInserterItems`
  and filter/sort by `frecency`.

Inserter changes:

- Both `inserter` and `inserter-with-shortcuts` have been changed to use
  `getInserterItems` and its new functionality.

noisysocks committed:
Memoize getInserterItems()

Use createSelector() to cache the results of getInserterItems() and
increase inserter performance.

tofumatt commented on editor/utils/test/allow-list.js:
```diff
@@ -0,0 +1,70 @@
+	} );
+
+	it( 'should parse a shorthand array', () => {
+		const allowList = parseAllowList( [ 'core/image', '!core/verse', '*' ] );
```
It seems like we're doing a lot of work here to avoid using a JS built-in: a plain object with the block name as a key and `true`/`false` as the allow value. Is this `'!core/verse'` type of argument common elsewhere? It seems to be extra code to run and test to avoid writing `{ 'core/image': false }` and instead write `[ '!core/image' ]`. It doesn't seem worth it to me.

The only benefit is maybe a list of allowed `parseAllowList( [ 'core/image', 'core/verse' ] ) => { 'core/image': true, 'core/verse': true }` but even then I say we just make objects required for where this is used and save a lot of code and make the API more consistent. Consistent APIs are good and lead to fewer bugs 😃

tofumatt replied:
Yeah, that makes sense. I'm cool with keeping just `[ 'core/image', 'core/verse' ]`; it's a nice shorthand for allowing a list of blocks.

tofumatt replied:
In terms of an easily understandable API I'd just make it another property. There's less room for error that way (it would be easier to spot what's allowed and what's not)... and actually I don't see why they couldn't be combined behind the scenes. Why not:


```js
<WhateverBlock
  allowedBlocks={ [ 'core/image', 'core/verse' ] }
/>
<WhateverBlock
  disallowedBlocks={ [ 'core/video' ] }
/>
/* This should throw an error, because when do you need an allow list and a disallow list when you don't set wildcards? */
<WhateverBlock
  allowedBlocks={ [ 'core/image', 'core/verse' ] }
  disallowedBlocks={ [ 'core/video' ] }
/>
```

That makes it way clearer and you don't have to manually set wildcards.

tofumatt replied:
That's how the components would see it... programmatically maybe `parseAllowList` is the wrong approach and it should be `allowList( [ 'core/image', 'core/verse' ] )` and `disallowedList( [ 'core/video' ] )`.

There's no logical need to allow both allowed and disallowed blocks to be explicitly defined in the list... right? 😄

noisysocks replied:
> There's no logical need to allow both allowed and disallowed blocks to be explicitly defined in the list... right? 😄

No, and that's part of why I like the object syntax. When typing `{ 'core/video': true }` you're forced to choose `true` or `false` 🙂

tofumatt replied:
The only issue with that is there is wildcard for either true or false after everything else, but the object API allows for explicit allowed and disallowed values. But with an allowed list approach, you don’t need to explicitly disallow any values with a `false` setting.

The object api allows folks to set things as both true and false which is confusing. It might be better to accept a flat array as either allowed Or disallowed, and then wildcard everything else as the opposite

noisysocks replied:
In a previous version of this PR I tried having only whitelists. The problem with, this, though was that you could get into situations where a parent and child conflict with one another. For example:

```js
// In foo/parent:
<InnerBlocks allowedBlocks={ [ 'foo/child-1' ] }>

// In foo/child-1:
allowedParents: []

// In foo/child-2:
allowedParents: [ 'foo/parent' ]
```

Here, it's not clear what should be allowed.

- Can `foo/child-1` be inserted into `foo/parent`? The parent says yes, but the child says no.
- Can `foo/child-2` be inserted into `foo/parent`? The parent says no, but the child says yes.

This is what motivated me to [introduce the object syntax](https://github.com/WordPress/gutenberg/issues/5540#issuecomment-386192518), which attempts to introduce a distinction between what a block _explicitly_ desires and what it _implicitly_ desires.

```js
// In foo/parent:
<InnerBlocks allowedBlocks={ { 'foo/child-1': true, '*': false } }>

// In foo/child-1:
allowedParents: { '*': false }

// In foo/child-2:
allowedParents: { 'foo/parent': true, '*': false }
```

Now, things are clearer.

- Can `foo/child-1` be inserted into `foo/parent`? The parent explicitly says yes, and the child implicily says no. Explicit wins, so the answer is yes.
- Can `foo/child-2` be inserted into `foo/parent`? The parent implicitly says no, and the child explicitly says yes. Explicit wins, so the answer is yes.

Being able to explicitly say no is useful, too. For example, `foo/parent` might want to prevent `foo/buggy` from ever being nested within it because it causes a nasty styling bug.

```js
// In foo/parent:
<InnerBlocks allowedBlocks={ { 'foo/child-1': true, 'foo/buggy': false, '*': false } }>
```

Here, it doesn't matter what `foo/buggy` specifies—it won't be allowed in `foo/parent`.

😅 Sorry about the long comment. I hope that explains some of my thinking behind the object syntax—let me know what you think.

tofumatt replied:
That makes sense... but it's very complex and I'd say tough to explain. You need to know whether something was explicitly/implicitly set to know if it should be allowed.

I find this approach unintuitive because I'd imagine different people will intuit different things from the settings.

Why not just err on the side of "any time something is not allowed: answer is no?" It's much simpler and means there isn't a complex formula to whether a block is allowed. If the parent or the child says no: answer is no.

The example:

```js
<InnerBlocks allowedBlocks={ { 'foo/child-1': true, 'foo/buggy': false, '*': false } }>
```

becomes:

```js
<InnerBlocks allowedBlocks={ [ 'foo/child-1' ] }>
```

As a new developer coming to this API I'd be confused why there are explicit and wildcard `false` values.

noisysocks replied:
> Why not just err on the side of "any time something is not allowed: answer is no?" 

Because if we err on the side of 'no', it means that a plugin developer cannot write a block that is insertable into a core block that uses `allowedBlocks` (e.g. [cover image](https://github.com/WordPress/gutenberg/pull/5452/files?utf8=✓&diff=unified#diff-30a5f912f6d7dc9db5d7c49abfced736R187)). This came up in https://github.com/WordPress/gutenberg/issues/5540#issuecomment-380816384.

I do concede that the object syntax is difficult to explain. I like it because of the flexibility it gives plugin developers, but I'd be happy to start with array whitelists (erring on the side of 'yes') and iterate based on developer feedback.

tofumatt commented on editor/store/selectors.js:
```diff
@@ -1354,28 +1408,16 @@ function getItemsFromInserts( state, inserts, allowedBlockTypes, maximum = MAX_R
-	return getItemsFromInserts( state, sortedInserts, allowedBlockTypes, maximum );
+	const items = getInserterItems( state, allowedBlockTypes );
+	const usefulItems = items.filter( ( item ) => item.utility > 0 );
+	const sortedItems = orderBy( usefulItems, [ 'utility', 'frecency' ], [ 'desc', 'desc' ] );
```
As mentioned, I think any reference to `'frecency'` should be annotated as it's not an extremely common concept in my experience.

noisysocks replied:
This function will be deleted in 3.0 so I won't add a comment here.

tofumatt commented on core-blocks/block/index.js:
```diff
@@ -173,7 +173,7 @@ export const name = 'core/block';
 	title: __( 'Shared Block' ),
 	category: 'shared',
-	isPrivate: true,
+	allowedParents: false,
```
Update: I've seen why this allows `true` and `false` types but I think they make the API unclear. I'd rather constants that signified the meaning of what `true` and `false` are achieving here instead of just using those values without info.

youknowriad commented on blocks/api/registration.js:
```diff
@@ -160,6 +161,14 @@ export function registerBlockType( name, settings ) {
+	if ( 'isPrivate' in settings ) {
+		deprecated( 'isPrivate', {
+			version: '3.0',
+			alternative: 'allowedPostTypes',
```
`isPrivate` means the block is available programmatically but not in the UI. When I read `allowedPostTypes` or `allowedParents`, it looks like in these case the blocks shouldn't be available entirely (even programmatically). So for me it seems that these two should be separate options? 

The difference is semantic for me, which means if I use isPrivate the block can be inserter everywhere without issue. While we could imagine adding a warning or an error if the block is used at a position that goes against its `allowedParents` config.

noisysocks replied:
`isPrivate: true` and `allowedParents: []` behave the same so my thought was to consolidate them. You're right that there's a slight semantic difference, though. I'm happy to add `isPrivate` back 🙂

cc. @gziolo

noisysocks replied:
`supports.inserting` kind of makes sense to me. I'm happy with that.

tofumatt reviewed (changes requested):
I have some issues with immediately understanding the code that are probably related to my being new to the project. Maybe just some comments as clarification would help there.

Beyond that, I think the list arguments are too fancy for little benefit. It means various blocks will use a slightly different API for declaring Allowed Parent Block Types making it hard to easily work on different blocks depending on who coded it and which API they prefer. I think it could use simple JS objects with `'*': true` and '*': false` as special values, rather than the much more flexible and less consistent argument types it takes now.

youknowriad commented on editor/store/selectors.js:
```diff
@@ -1204,141 +1229,170 @@ export function getNotices( state ) {
-	if ( ! allowedBlockTypes || ! sharedBlock ) {
-		return null;
-	}
+		const calculateUtility = ( category, count, isContextual ) => {
```
If I understand properly the utility is essentially usefull when displaying the "suggested" category right?

Order the items in the inserter: Suggested > frecent > common

And it's also used to exclude items from the Suggestions (utility = 0).

I think it's a bit confusing (maybe just because it's a new API), what if we don't filter at all, if we splice later (seems like we do), we don't need to filter.

noisysocks replied:
Yep, the behaviour of the _Suggested_ tab is now to show items in the following order:

1. Blocks that are contextually useful. For example, if you are inserting into a _Product_ block, _Add to Cart_ is contextually useful.
2. Blocks that have been inserted by the current user in the past.
3. Blocks that are in the common category.
4. All other blocks.

I introduced `utility` as a way to achieve this, since the inserter logic was growing quite complex, e.g. `fillWithCommonBlocks` was quite difficult to reason about.

youknowriad commented on editor/store/selectors.js:
```diff
@@ -1204,141 +1229,170 @@ export function getNotices( state ) {
+ *                                        ordering items by relevancy.
+ */
+export const getInserterItems = createSelector(
+	( state, editorAllowedBlockTypes, parentUID = null ) => {
```
Hopefully with #6500 we'd be able to remove the second arg entirely.

youknowriad commented on editor/utils/allow-list.js:
```diff
@@ -0,0 +1,67 @@
+ * Some examples:
+ *
+ * { 'core/image': true, '*': false } specifies that only images are allowed.
+ * { 'core/verse': false, '*': true } specifies that everything except verses are allowed.
```
Do we need to support the block lists? I'm asking cause it seems redundant with the `allowedBlocks` prop of the `InnerBlock` component. and in that case we could simplify the API here and just provider a list of allowed parents.

noisysocks replied:
Do you mean allow the advanced syntax (`{ 'foo': false, 'bar': true, '*': true }`) in `allowedBlocks` but only simple arrays (`[ 'foo', 'bar' ]`) in `allowedParents`?

My preference is that we support the same syntax in `allowed_block_types` (the PHP filter), `allowedBlocks` (the `InnerBlocks` prop), and `allowedParents` since it's no extra work for us and introduces some consistency throughout the entire blocks API.

youknowriad replied:
Oh sorry, what I meant is that we should only support `[ 'foo', 'bar' ]` in all of these, as it's simple and I believe we can achieve everything with it? unless I'm missing a use-case?

tofumatt replied:
Doing exceptions is harder with the array-only list (how do you make `({ 'foo': false, 'bar': true, '*': true })`?), so unless we keep the `['!foo', 'bar', '*']` option around (which is more code/complexity and less understandable as plain JS) we'd need to allow objects as arguments.

youknowriad replied:
Do we really need to do exceptions? What use-cases can't be achieved with a combination of `allowParents` and `allowedBlocks` (which is `allowChildren`)?

gziolo replied:
Actually we need to cover two cases which are:
- list all allowed blocks
- list all disallowed blocks, which is in other words, all but listed blocks

noisysocks replied:
> What use-cases can't be achieved with a combination of allowParents and allowedBlocks (which is allowChildren)?
 
A plugin developer wouldn't be able to e.g. create a block that allows all children except for `core/image` because they can't change the `allowedParents` of an already registered block.

> Do we really need to do exceptions?

I mean—no 😛 But my thinking was that it could be useful for plugin developers.

In the interest of arriving at a consensus I'd be happy to start with only whitelist arrays. We can expand to allowing whitelist+blacklist objects in the future based on feedback we get from plugin developers.

Let me know what y'all think.

youknowriad replied:
I think you convinced me we need both, particularly this argument 👍 

> A plugin developer wouldn't be able to e.g. create a block that allows all children except for core/image because they can't change the allowedParents of an already registered block.

youknowriad reviewed (commented):
I personally don't like the ambiguity between post types and blocks :) what if I want a block to be allowed in a parent block `plugin/product` but not the CPT `plugin/product`. Granted that it's not too common but I still prefer explicitness.

noisysocks commented on editor/store/selectors.js:
```diff
@@ -1193,8 +1197,29 @@ export function getNotices( state ) {
+ * items (e.g. a regular block type) and dynamic items (e.g. a shared block).
+ *
+ * @param {Object}           state                   Global application state.
+ * @param {string[]|boolean} editorAllowedBlockTypes Allowed block types, or true/false to
```
😅 I agree that using `true` and `false` for this is regrettable. Happy to move away from it. We'd have to `deprecate()` as this API is public.

tofumatt replied:
Ah, okay. Deprecating it would be 👍  😄

noisysocks commented on editor/components/inserter-with-shortcuts/index.js:
```diff
@@ -17,18 +17,30 @@ import { withDispatch, withSelect } from '@wordpress/data';
+}
+
+function orderItems( items ) {
+	return orderBy( items, [ 'utility', 'frecency' ], [ 'desc', 'desc' ] );
```
Agreed—I wrote about this [here](https://github.com/WordPress/gutenberg/pull/6346#discussion_r186585878) but it could definitely use a comment.

noisysocks replied:
Added a comment in 0869dec79!

noisysocks commented:
Thanks for the thoughts @tofumatt and @youknowriad!

> Beyond that, I think the list arguments are too fancy for little benefit. It means various blocks will use a slightly different API for declaring Allowed Parent Block Types making it hard to easily work on different blocks depending on who coded it and which API they prefer. I think it could use simple JS objects with '*': true and '*': false` as special values, rather than the much more flexible and less consistent argument types it takes now.

You're right that it's overkill. I'm thinking let's only support arrays and objects, e.g.:

```js
parseAllowList( [] ) => { '*': false }
parseAllowList( [ 'core/image' ] ) => { 'core/image': true, '*': false }
parseAllowList( [ 'core/image': false, '*': true ] ) => { 'core/image': false, '*': true }
```

I think this is a good balance between plugin developer ergonomics and Gutenberg complexity.

Booleans would also be supported for a short while with a console warning that tells folks to replace `allowedBlocks={ false }` with `allowedBlocks={ false }` and `allowedBlocks={ true }` with ``.

> I personally don't like the ambiguity between post types and blocks :) what if I want a block to be allowed in a parent block plugin/product but not the CPT plugin/product. Granted that it's not too common but I still prefer explicitness.

I think @mtias was the only one pushing for this. I'm happy to go either way on it 🙂

noisysocks committed:
Test that the inserter limits the suggested items to MAX_SUGGESTED_ITEMS

noisysocks committed:
Add comments explaining the sort order of inserter items

noisysocks commented on editor/components/inserter/menu.js:
```diff
@@ -144,7 +147,8 @@ export class InserterMenu extends Component {
 	sortItems( items ) {
 		if ( 'suggested' === this.state.tab && ! this.state.filterValue ) {
-			return items;
+			const sortedItems = orderBy( items, [ 'utility', 'frecency' ], [ 'desc', 'desc' ] );
```
Added a comment in 0869dec79!

noisysocks commented on editor/components/inserter/menu.js:
```diff
@@ -144,7 +147,8 @@ export class InserterMenu extends Component {
 		if ( 'suggested' === this.state.tab && ! this.state.filterValue ) {
-			return items;
+			const sortedItems = orderBy( items, [ 'utility', 'frecency' ], [ 'desc', 'desc' ] );
+			return sortedItems.slice( 0, MAX_SUGGESTED_ITEMS );
```
Added a test in cd3e91e8f!

tofumatt reviewed (commented):
Definitely a fan of the simpler, consolidated API for `allowedBlocks={}` 👍

gziolo commented:
I’m all for discontinuing  support for boolean values. We also need to keep im mind that we need to pick a schema which can be expressed in both PHP and JS, because we use this filter on the server which allows to pick which blocks can be inserted. At the moment we support only whitelist approach on the server, but theres is expectation to provide a way to support all blocks but those listed. I personally would seek for solution which is flexible enough to prevent us from introducing another property.

noisysocks commented:
Quick update on this, since there's a few disparate conversations going on in threads:

I'm aiming to get a consensus on the API before I continue any further. We're also waiting on some thoughts from @mtias who said he'll pop in soon.

I think we all agree on:

- Deprecating booleans e.g. `allowedBlocks={ false }` and `add_filter( 'allowed_block_types', function() { return false; } )` because it's confusing and specifying e.g. an empty array is more meaningful
- Rename `isPrivate` to something like `supports.insertion` instead of deprecating it

Still contentious:

- Should `allowedBlocks`, `allowedParents` and `allowed_block_types` only support whitelisting blocks or should we support both whitelisting and blacklisting? I'm fond of the object syntax in this PR that allows both whitelisting and blacklisting. Very happy to start with only whitelist arrays for now, though, and iterate on this.
- Should `allowedParents` be used for post types _as well as_ block types? e.g. `allowedParents: [ 'post', 'page', 'plugin/cool-block' ]` Or should a separate property be introduced? e.g. `allowedParentBlocks: [ 'plugin/cool-block' ], allowedPostTypes: [ 'post', 'page' ]`

mtias commented:
Let's punt the decision on supporting post types under the same API so that it doesn't block us from merging this.

noisysocks commented:
Thanks for the thoughts, y'all. Going to close this and split something small and mergeable from it.
